### DISCUSSION TITLE
[SDK] Can't get the signer for zkSync in web3-react

### DISCUSSION THREAD

**Discussion by aristidesgpg** [zkSync-Community-Hub/zksync-developers#48] [SDKs] [answered]
### Environment

Testnet

### Acknowledgement

- [ ] I understand that issues are expected with the Python and Java SDK as they are being updated.

### Issue Description

- issue
  can't get the signer for zkSync.
  I need to call the write function of the smart contract and tried to get the signer for zkSync, but it has an error.
![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/48382274/d555111e-3d10-4899-a318-3c3a74be0ddc)

After signing, it has an error.
![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/48382274/cd393d03-942b-48ea-9ff7-fb0172008f8b)

- Environment
   Windows, Metamask, Node 16.  Node 18
- Libraries and framework
   Next.js, Web3-react, zksync

### Expected Behavior

Should get the zkSync signer.

### Code Example

import * as zksync from 'zksync';

const { library } = useWeb3React();
const syncProvider = await zksync.getDefaultProvider('goerli');  
const syncWallet = await zksync.Wallet.fromEthSigner(library.getSigner(), syncProvider);  // This is causing the error.  It displays the 
                                  //Metamask popup to sign the signature.  After clicking the sign button, it throws the error.

**Comment by dutterbutter** [comment]
Hello 👋,

Thank you so much for bringing this issue to our attention! 🙏 

Your detailed report, complete with screenshots and code snippets, is very helpful! 👏📸

To further assist you, it would be beneficial if I could reproduce the issue on my end. Could you please provide a repo with the minimal code that reproduces this issue? This will help me pinpoint the problem and provide a solution more effectively.

**Reply by aristidesgpg** [reply]
@dutterbutter Thanks for your response.
Unfortunately, I can't share our repo, but I already shared my full code.

**Reply by aristidesgpg** [reply]
Or if you have any examples that is using Web3 React, could you please share it with me?

**Reply by dutterbutter** [reply]
You have the option to set up a wallet without resorting to Web3 React. For a comprehensive guide, I suggest that you refer to our `zksync-web3` documentation. You can access it directly from the following link:👉  [zksync-web3 documentation](https://era.zksync.io/docs/api/js/getting-started.html).

For us to provide a more targeted and effective solution to your problem, it would be immensely helpful if you could share a minimal working example of the issue you're experiencing. This will allow us to replicate the issue on our end and assist you more effectively.

**Reply by aristidesgpg** [reply]
@dutterbutter Does your SDK 'zksync-web3' will display the connect popup? I can see it requires the private key.

**Reply by aristidesgpg** [reply]
You can just install zksync and web3-react libraries and run the below code.

import * as zksync from 'zksync';

const { library } = useWeb3React();
const syncProvider = await zksync.getDefaultProvider('goerli');
const syncWallet = await zksync.Wallet.fromEthSigner(library.getSigner(), syncProvider); // This is causing the error. It displays the
//Metamask popup to sign the signature. After clicking the sign button, it throws the error.

**Reply by aristidesgpg** [reply]
@dutterbutter I created the repo so that you can test and invited you.

**Reply by dutterbutter** [reply]
Hello, I reviewed the minimal example provided and have provided a few takeaways:

1. You are referencing a library called `zksync`, you should only make use of `zksync-web3` the official javascript sdk. Thus your imports should be changed to:
```
import { Web3Provider, <OTHER CLASSES TO IMPORT (e.g. Wallet, ContractFactory> } from "zksync-web3";
```
2. Some of the methods you're trying to invoke (like `zksync.Wallet.fromEthSigner`) aren't recognized. If your goal is to interface with a MetaMask account, consider this approach:
```
import { Web3Provider } from "zksync-web3";

const provider = new Web3Provider((window as any).ethereum);
await provider.send("eth_requestAccounts", []);   
const signerInstance = provider.getSigner();
```

This will give you the signer that you can then pass to your ContractFactory. 

3. Be cautious about supplying the appropriate number of constructor arguments when deploying your contract. I observed some inconsistency in your code:
```
await TokenFactory.deploy(); // this line will fail as it expects 4 constructor arguments
await TokenFactory.deploy(name, 'symbol', parseEther('123123123'), true);
```
Ensure you're invoking the deploy method with the right number of arguments that align with your contract's constructor.

**Reply by aristidesgpg** [reply]
Thanks for checking.
This repo is just for demo purposes, so there may be some wrong codes. Sorry about that.

Btw, can you check again? I updated the code. It fails to deploy the contract.
And we also need to support the wallet connect.

**Comment by aristidesgpg** [comment]
@dutterbutter If you have any issues while running the repo locally, please let me know.

**Reply by aristidesgpg** [reply]
Now I am getting the below error.  I think it's same as https://github.com/wighawag/hardhat-deploy/issues/453.
![image](https://github.com/zkSync-Community-Hub/zkync-developers/assets/48382274/c11d9f45-291b-4957-b365-34f113f4316f)

**Reply by uF4No** [reply, marked as answer]
Make sure your contract is compiled with zksolc. You can find a hardhat.config example here: https://era.zksync.io/docs/tools/hardhat/migrating-to-zksync.html

**Reply by aristidesgpg** [reply]
@uF4No Awesome!
It's working.
Btw, will it be the same with other chains to verify the contract?  Or are there any additional steps?

**Reply by uF4No** [reply]
Verification is pretty similar. 
If your contract is simple, you can [verify directly from the explorer](https://era.zksync.io/docs/tools/block-explorer/contract-verification.html).
If you prefer to do it from the terminal or programatically, you need to install our [verification plugin](https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-verify.html), import it in the hardhat.config.ts file and run the `verify:verify` task. All the details are [in our docs](https://era.zksync.io/docs/tools/hardhat/hardhat-zksync-verify.html#commands) 🤙
